app/agent/block-registry/registry.service.ts

```diff
@@ -0,0 +1,1048 @@
+  async expose(
+    what: ImageId | ContainerId,
+    method: EXPOSE_METHOD,
+  ): Promise<{ device: string; readonly: boolean } | { mountPoint: string; readonly: boolean }> {
+    if (!what.startsWith("im_") && !what.startsWith("ct_")) {
+      throw new Error(`Invalid id: ${what}`);
+    }
+    const readonly = what.startsWith("im_");
+    switch (method) {
+      case EXPOSE_METHOD.HOST_MOUNT: {
+        const bd = await this.expose(what, EXPOSE_METHOD.BLOCK_DEV);
+        const mountPoint = path.join(this.paths.mounts, what);
+        await fs.mkdir(mountPoint).catch(catchAlreadyExists);
+        try {
+          let flags = readonly ? ["--read-only"] : ["--read-write", "-o", "discard"];
+          const t1 = performance.now();
+          // TODO: Calling the mount syscall directly is much much more faster
+          await execCmdAsync("mount", ...flags, bd.device, mountPoint);
+          this.logger.info(`mount for ${what} took: ${(performance.now() - t1).toFixed(2)} ms`);
+        } catch (err) {
+          if (!(err as Error).message.includes("already mounted")) throw err;
+        }
+        return { mountPoint, readonly };
+      }
+      case EXPOSE_METHOD.BLOCK_DEV: {
+        const t1 = performance.now();
+        // Get the tcmuSubtype
+        const tcmuSubtype = await this.getTCMUSubtype();
+        const tcmuStorageObjectId = `${tcmuSubtype}_${what}`;
+        const tcmuPath = path.join(this.paths.tcmuHBA, tcmuStorageObjectId);
+        // Create a storage object on that HBA, if it already exists then check for a LUN
+        let aluaMembers: string[] = [];
+        await fs.mkdir(tcmuPath).catch(async (err: Error) => {
+          if (!err.message.startsWith("EEXIST")) throw err;
+          // Ok check for LUN
+          aluaMembers = await this.getTCMUAluaMembers(tcmuStorageObjectId);
+        });
+        // Check if the device was not mapped to a LUN
+        if (aluaMembers.length === 0) {
+          // WARNING: The upstream kernel is borked and data corruption occurs when concurrent enable requests are made
+          //          There is no mutex in the kernel protecting concurrent enable requests
+          //          Due to this get an exclusive flock on the enable file
+          //          The fs-ext library is unmaintained and will stall the event loop when there are more than 3 concurrent flocks
+          //          waiting, due to this we request a flock in unblocking mode and sleep for some time.
+          //          TODO: Consider writing a rust addon for flocks cause there is no good library for this in node .-.
+          const enableFd = await fs.open(path.join(tcmuPath, "enable"), "w");
+          let tcmuAttachFailed = false;
+          try {
+            let flockAcquired = false;
+            while (!flockAcquired) {
+              try {
+                await new Promise((resolve, reject) => {
+                  flock(enableFd.fd, "exnb", (err) => {
```

Comment:
you can use promisify instead: https://nodejs.org/api/util.html#utilpromisifyoriginal

Reply:
Just tried that and it doesn't work - `promisify` expects an `(err, value)` callback, where the flock library uses an `(err_or_null)` callback. Additionally promisify converted the `(number, string, callback)` function signature into a `(number, numer)` type sigrature.